src/util/invariant.h

```diff
@@ -81,25 +132,79 @@ class invariant_failedt: public std::logic_error
 
 
 #else
-// CPROVER_INVARIANT_PRINT_STACK_TRACE affects the implementation of
-// this function but not it's generation from the macro
 
-void check_invariant(
+void print_backtrace(
+  const char * const file,
+  const char * const function,
+  const int line,
+  std::ostream &out);
+
+std::string get_backtrace(
+  const char * const file,
+  const char * const function,
+  const int line);
+
+/// Takes a backtrace, gives it to the reason structure, then aborts, printing
+/// reason.what() (which therefore includes the backtrace).
+/// In future this may throw `reason` instead of aborting.
+/// \param file : C string giving the name of the file.
+/// \param function : C string giving the name of the function.
+/// \param line : The line number of the invariant
+/// \param reason : (derived from invariant_failedt) error report structure;
+///  its backtrace member will be set before it is used.
+template<class T>
+typename std::enable_if<std::is_base_of<invariant_failedt, T>::value>::type
+invariant_violated(
   const char * const file,
   const char * const function,
   const int line,
-  const bool condition,
-  const char * const reason);
+  T reason)
+{
+  reason.set_backtrace(get_backtrace(file, function, line));
+  // We now have a structured exception ready to use;
+  // in future this is the place to put a 'throw'.
+  std::cerr << "--- begin invariant violation report ---\n";
+  std::cerr << reason.what() << '\n';
+  std::cerr << "--- end invariant violation report ---\n";
+  abort();
+}
+
+/// Takes a backtrace, constructs an invariant_violatedt from reason and the
+/// backtrace, aborts printing the invariant's description.
+/// In future this may throw rather than aborting.
+/// \param file : C string giving the name of the file.
+/// \param function : C string giving the name of the function.
+/// \param line : The line number of the invariant
+/// \param reason : (std::string or const char*) brief description of the invariant
+///   violation.
+template<class T>
+typename std::enable_if<std::is_constructible<std::string, T>::value>::type
+invariant_violated(
+  const char * const file,
+  const char * const function,
+  const int line,
+  const T &reason)
+{
+  invariant_failedt failure(reason);
+  invariant_violated(file, function, line, failure);
+}
 
 #ifdef _MSC_VER
 #define INVARIANT(CONDITION, REASON) \
-  check_invariant(__FILE__, __FUNCTION__, __LINE__, (CONDITION), (REASON))
+  do \
+  { \
+  if(!(CONDITION)) \
+    invariant_violated(__FILE__, __FUNCTION__, __LINE__, (REASON)); \
+  } while(0)
```

Comment:
Presumably this is to make a statement that requires a trailing semicolon to make INVARIANT behave syntactically as much like a normal function as possible. If so, a comment saying so would be helpful.